core/platform/auth/firebase_auth_services.py

```diff
@@ -393,3 +371,136 @@ def associate_multi_auth_ids_with_user_ids(auth_id_user_id_pairs):
+    user's experience. If Firebase could not complete an operation, then that
+    simply means that authentication has failed, there's no need to provide a
+    user-facing error.
+
+    Yields:
+        None. No relevent context expression.
+    """
+    try:
+        yield
+    except (ValueError, firebase_exceptions.FirebaseError):
+        logging.exception('Firebase Admin SDK raised an exception!')
+
+
+@contextlib.contextmanager
+def _firebase_admin_context():
+    """Returns a context for calling the Firebase Admin SDK.
+
+    Yields:
+        None. No relevent context expression.
+    """
+    try:
+        app = firebase_admin.initialize_app(
+            options={'projectId': feconf.OPPIA_PROJECT_ID})
+        yield
+    except (ValueError, firebase_exceptions.FirebaseError):
+        logging.exception('Firebase Admin SDK raised an exception!')
+    finally:
+        if app is not None:
+            firebase_admin.delete_app(app)
+
+
+def _get_session_cookie(request):
+    """Returns the session cookie authorizing the signed in user, if present.
+
+    Args:
+        request: webapp2.Request. The HTTP request to inspect.
+
+    Returns:
+        str|None. Value of the session cookie authorizing the signed in user, if
+        present, otherwise None.
+    """
+    return request.cookies.get(feconf.FIREBASE_SESSION_COOKIE_NAME)
+
+
+def _get_id_token(request):
+    """Returns the ID token authorizing a user, or None if missing.
+
+    Oppia uses the OAuth 2.0's Bearer authentication scheme to send ID Tokens.
+
+    Bearer authentication (a.k.a. token authentication) is an HTTP
+    authentication scheme based on "bearer tokens", an encrypted JWT generated
+    by a trusted identity provider in response to login requests.
+
+    The name "Bearer authentication" can be understood as: "give access to the
+    bearer of this token." These tokens _must_ be sent in the `Authorization`
+    header of HTTP requests, and _must_ have the format: `Bearer <token>`.
+
+    Learn more about:
+        HTTP authentication schemes:
+            https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication
+        OAuth 2.0 Bearer authentication scheme:
+            https://oauth.net/2/bearer-tokens/
+        OpenID Connect 1.0 ID Tokens:
+            https://openid.net/specs/openid-connect-core-1_0.html#IDToken
+
+    Args:
+        request: webapp2.Request. The HTTP request to inspect.
+
+    Returns:
+        str|None. The ID Token of the request, if present, otherwise None.
+    """
+    scheme, _, token = request.headers.get('Authorization', '').partition(' ')
+    return token if scheme == 'Bearer' else None
+
+
+def _get_auth_claims_from_id_token(token):
+    """Returns claims from the ID Token, or None if invalid.
+
+    Args:
+        token: str|None. The token to extract claims from.
+
+    Returns:
+        auth_domain.AuthClaims|None. The claims from the ID Token, if available.
```

Comment:
Optional: totally fine to just write `AuthClaims|None` if you prefer, unless there's a need to dissambiguate. Ditto below.

Reply:
Done.